lib/cog/bootstrap.ex

```diff
@@ -27,52 +32,65 @@ defmodule Cog.Bootstrap do
   Create a user with permissions in the embedded namespace then
   returns the admin user
   """
-  def bootstrap do
-    {:ok, user} = Repo.transaction(fn() ->
-      user = create_admin
-      grant_embedded_permissions_to(user)
-      create_namespace(@site_namespace)
-      user
-    end)
+  def bootstrap,
+  do: bootstrap(@default_admin_params)
 
-    {:ok, user}
-  end
+  def bootstrap(params) when params == %{},
+  do: bootstrap(@default_admin_params)
 
-  # TODO: require either username or email (or both), but if one is
-  #       missing, that's  OK.
-  # TODO: Make first/last name optional?
+  def bootstrap(params) do
+    Repo.transaction(fn() ->
+      user = create_admin(params)
+      role = create_by_name(Role, Cog.admin_role)
+      group = create_by_name(Group, Cog.admin_group)
 
-  # Create a bootstrap admin user. The username will always be
-  # `@admin_name`, and the password will be randomly-generated, which
-  # will be available in the `password` field of the returned
-  # `Cog.Models.User` struct.
-  defp create_admin do
-    # Horrible quick n' dirty hack to strip
-    # ; and # from passwords so ConfigParse_Ex doesn't
-    # choke on them.
-    password = String.replace(Cog.Passwords.generate_password(32), ~r/[;#]/, "")
+      grant_embedded_permissions_to(role)
+      grant_role_to_group(group, role)
+      add_user_to_group(user, group)
 
-    params = %{
-      username: @admin_name,
-      first_name: "Cog",
-      last_name: "Administrator",
-      password: password,
-      email_address: "cog@localhost"}
+      create_by_name(Namespace, Cog.site_namespace)
+
+      user
+    end)
+  end
 
+  # Create a bootstrap admin user from the given parameter map. If
+  # the password is empty, generate a random one. Returns the username
+  # and password in the response.
+  defp create_admin(params) do
+    params = case Map.get(params, "password") do
+               nil ->
+                 Map.put(params, "password", generate_safe_password)
+               _ ->
+                 params
+             end
```

Comment:
Totally not necessary, but might be nice to use `Map.put_new` or `Map.put_new_lazy` here:

```elixir
Map.put_new_lazy(params, "password", &generate_safe_password/0)
```